homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,326 @@
+                                               data[ATTR_SERVICE_DATA],
+                                               True,
+                                               context(request))
+            except (vol.Invalid, ServiceNotFound):
+                raise HTTPBadRequest()
+
+        return json_response(changed_states)
+    elif webhook_type == WEBHOOK_TYPE_FIRE_EVENT:
+        event_type = data[ATTR_EVENT_TYPE]
+        hass.bus.fire(event_type, data[ATTR_EVENT_DATA], ha.EventOrigin.remote,
+                      context(request))
+        return json_response({"message": "Event {} fired.".format(event_type)})
+    elif webhook_type == WEBHOOK_TYPE_RENDER_TEMPLATE:
+        try:
+            tpl = template.Template(data[ATTR_TEMPLATE], hass)
+            rendered = tpl.async_render(data.get(ATTR_TEMPLATE_VARIABLES))
+            return json_response({"rendered": rendered})
+        except (ValueError, TemplateError) as ex:
+            return json_response("Error rendering template: {}".format(ex),
+                                 status=HTTP_BAD_REQUEST)
+
+    elif webhook_type == WEBHOOK_TYPE_UPDATE_LOCATION:
+        await hass.services.async_call(DEVICE_TRACKER_DOMAIN,
+                                       DEVICE_TRACKER_SEE, data)
+        return json_response([])
+
+
+def supports_encryption():
+    """Test if we support encryption."""
+    try:
+        import nacl   # noqa pylint: disable=unused-import
+        return True
+    except OSError:
+        return False
+
+
+def device_for_webhook_id(hass, webhook_id):
+    """Return the device name for the webhook ID."""
+    for device_name, device in hass.data[DOMAIN].items():
+        if device.get(CONF_WEBHOOK_ID) == webhook_id:
+            return device
+    return None
+
+
+async def async_setup(hass, config):
+    """Set up the mobile app component."""
+    conf = config.get(DOMAIN)
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    app_config = await store.async_load()
+    if app_config is None:
+        app_config = {}
+
+    hass.data[DOMAIN] = app_config
+
+    for name, device in app_config.items():
+        if CONF_WEBHOOK_ID in device:
+            webhook.async_register(hass, DOMAIN, 'Mobile app',
+                                   device[CONF_WEBHOOK_ID], handle_webhook)
+
+    if conf is not None:
+        hass.async_create_task(hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up an mobile app entry."""
+    hass.http.register_view(RegisterDeviceView())
+
+    return True
+
+
+class RegisterDeviceView(HomeAssistantView):
+    """A view that accepts device registration requests."""
+
+    url = '/api/mobile_app/register'
+    name = 'api:mobile_app:register'
+
+    def __init__(self):
+        """Initialize the view."""
+
+    async def post(self, request):
```

Comment:
We've got a helper for this:

https://github.com/home-assistant/home-assistant/blob/229d19bb20f4fc318c784517c667b92ca99f38fd/homeassistant/components/hassio/auth.py#L46-L47